pkg/messaging/direct_messaging.go

```diff
@@ -208,9 +234,14 @@ func (d *directMessaging) invokeWithRetry(
 			code := status.Code(rErr)
 			if code == codes.Unavailable || code == codes.Unauthenticated {
 				// Destroy the connection and force a re-connection on the next attempt
+				// We also remove the resolved name from the cache
 				teardown(true)
+				if app.cacheKey != "" && d.resolverCache != nil {
+					d.resolverCache.Delete(app.cacheKey)
```

Comment:
Deletion now occurs when unavailable which is correct, but also applies to unauthenticated. I believe we should only remove if unavailable. For example, an unauthenticated message will happen during a certificate rollover of the target sidecar. In this case, the endpoint is there but the calls should be retried until it succeeds (when the target sidecar's server has a valid cert)

Reply:
I think deletion from cache when un-authenticated is fine. This will cover the case of an app that is replacing a different app at the same IP.

Cert rollovers _shouldn't_ cause errors, because we rotate certs before their validity ends. As long as they are signed by the same CA, callers should continue to accept them with 0 downtime.

I would also err on the side of caution personally and over-purging from the cache, than potentially keeping stale values.